### ISSUE TITLE
Any plans to expose presolve and postsolve to godot userspace?

### ISSUE THREAD

**Issue by jordo** [briansemrau/godot_box2d#13]
Box2DWorld currently implements the contact listener presolve and postsolve.  Any thoughts about exposing this functionality to godot?

**Comment by briansemrau** [comment]
I believe that could be useful for advanced features like one-way platforms by disabling contacts, right? Yeah, I'll consider this a missing feature.

**Comment by jordo** [comment]
We always end up doing a lot of different stuff in presolve.  Modify restitution on the fly, disable the contact, one-way contacts, adjust the specific impulse depending on some game state, etc. 

It's pretty powerful from a game design standpoint, and definitely a plus to expose this if it's available.  Its something I miss dearly in godot. 

Post solve not as much, but we sometimes do stuff there too.

**Comment by jordo** [comment]
There's some added discussion to be had here, because supporting a PreSolve callback has some interesting constraints and potential ramifications for future enhancements such as multi-threading and moving any stepping of the world into the godot physics server.

The biggest one, off hand, and first one to tackle as a design decision is the access to any function like (SetTransform, or creating bodies or destroying bodies) in the callback itself.  Within the presolve callback the b2world is locked, which means that a whole schwack of functions will just not work inside this callback  (there are assertions when Box2D is built with debug on), check for world->IsLocked().

This is a pretty common starting occurance for developers who use the PreSolve callback.  They will end up setting a body's transfom midway through the solver, (which Box2D will bark at you).  I think there's a potential elegant solution with how this can be integrated into a godot front end, but will think some more on it.

Things to discuss are going to be the best way to handle this currently?  And further on with how to handle this when step() execution is potentially asynchronous with a scene update.

**Comment by briansemrau** [comment]
For implementation, I currently believe the most sensible way would be to add the pre/postsolve callback as a signal emitted from the Box2DWorld node (though I'm open to suggestion).

In Godot 2D physics it's not unusual to have thread-based restrictions in certain functions. For example, accessing the space should only be done from `_physics_process` ([docs](https://docs.godotengine.org/en/stable/tutorials/physics/ray-casting.html#accessing-space)). It might be sufficient to indicate in documentation that any actions from within these signal callbacks must use `call_deferred`, and that physics state changes must happen only within _physics_process.
From my understanding (might need testing), Area2D/PhysicsBody2D signal callbacks for `body_entered`, `body_exited`, etc. are all handled immediately on the physics thread.

I believe if we imitate the current solution Godot has for asynchronous physics, we should be okay. The future implementation for this will likely be creating a Box2DPhysicsServer. In Physics2DServer there is a fail condition for getting the direct space/body state from the wrong thread.

**Comment by jordo** [comment]
hmmm, i was thinking the presolve and post solve should follow the same signal emission point as the other callbacks "body_entered", etc (signal from the bodies).  They are effectively dealing with the same collision.

**Comment by jordo** [comment]
i guess my general thoughts are the less footguns the possible.   i.e. if there's a way to do things that don't require reading the documentation the better.  For the callback issues in pre-solve, it's mostly about accessing functions in which at the time will be locked by the b2world.

For this issue, even modifying a transform in _physics_process, but within a pre-solve context will be problematic as the world will be locked

**Comment by briansemrau** [comment]
What do you think about putting all these state modification actions (transform, create, delete, etc.) into a queue whenever the world is locked that later gets handled by the server between steps? It would be a lot of work, but creating the server is a very similar effort (see how the multithreaded physics server is implemented [here](https://github.com/godotengine/godot/blob/master/servers/physics_2d/physics_server_2d_wrap_mt.h)).

Of course, that would be for when the box2d physics server is implemented, which doesn't solve the issue now

**Comment by briansemrau** [comment]
> 
> 
> hmmm, i was thinking the presolve and post solve should follow the same signal emission point as the other callbacks "body_entered", etc (signal from the bodies). They are effectively dealing with the same collision.

Yeah, this is true. I think the best argument for not totally replacing those signals is just to keep feature parity with Godot.

**Comment by jordo** [comment]
I guess its up to discuss how close to feature parity we want, there's pros and cons.  Sure "body_entered" is the same as BeginContact(), but the nice thing about using something thats 1:1 with Box2D api is that you can just read the Box2D docs and understand the callback/signals.

The pro with mapping, is it's easy to move from godot 2d physics to box2d (same api, with potential addons).  The downside is you can't look at box2d documentation and just see that begincontact maps to "begincontact"... etc

**Comment by jordo** [comment]
Regarding the state modifications during solve (pre-solve), ya of course the application level solution to this is to queue those events and process them post step().  (This is what we do, and probably what everyone does).  But that doesn't explicity stop someone from calling `$ballbody.position = x` directly inside the pre-solve, which will break.

I have some thoughts on how to solve this elegantly, but you can't queue the actual pre-solve callbacks for later.  The whole point of pre-solve is to intercept the solver itself during a step().  (In which the world lists are locked (b2body list and b2contact list), and you can't change a linked list while iterating them.  SetTransform() ends up finding new contacts, which changes the internal b2contact list, which is why it's not allowed to be called during the solve as well.

**Comment by jordo** [comment]
The nice thing, is we will know when the world is locked, and we know what functions will work or what won't.  There's a good chance some local state in each Box2DBody node, and a dirty bit can be used to wrap a callback in post step().  Same thing with creating bodies, we can create the body, and schedule additional to the world post step().  I've give some more thought to this, but I wanna port my project first and then play around with pre-solve.

**Comment by jordo** [comment]
But ya, in general, if we're going to emit signals, we should stay consistent.  If coming from the bodies, it can be "body entered", "pre solve", etc.

Actually, there's a good reason signals should just come from both (come from the world and body nodes).  Leave it up to the user if they receive from the bodys or the world node.  I actually prefer signal handling from the world, because then our game logic isn't spread over a whole bunch of different nodes.    The callback can inspect the two different body node arguments, and you don't end up with the same signal emitting on both collision nodes.

**Comment by briansemrau** [comment]
Because of the proximity of code, should I address #6 now or hold off until after you look at making changes with presolve?

**Comment by jordo** [comment]
you should be fine to continue on #6

**Comment by jordo** [comment]
starting some work and investigations on this one
